Add 19 additional languages

David-Maisonave opened the pull request:
This pull request is only for LanguageDictionary.cs and AggregateTranslator.cs.
I'm sorry, but I'm not sure how to remove the other two files from the Pull Request.

Changes to the LanguageDictionary.cs include the following:
* 19 Additional languages
* Language name corrections IAW [ISO 639-1](https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes)
* Class to allow external exposure to the language list.
* The language list is sorted, so as to more easily identify any duplicates, and it makes it easier to find a language on the list.

I added an extra try catch on AggregateTranslator.cs, because in some cases an exception was thrown inside the catch code.

FYI:
I changed the csproj file, because it didn't compile with my project while it had the reference to .net 2.x on it. Feel free to ignore those changes.

David-Maisonave committed:
Merged code from ABetterTranslator

David-Maisonave committed:
Update README.md

David-Maisonave committed:
Update README.md

David-Maisonave committed:
Update README.md

David-Maisonave committed:
Update LanguageDictionary.cs

d4n3436 commented:
Hello, and thanks for contributing.
Could you please edit the code so the only changes are made to the `_languages` dictionary? The extra comments are unnecessary too. Also, you can undo the changes to the other files through your code editor.

> 19 Additional languages

Could you please provide your sources for these new languages? I obtained the languages from official documentation and inspecting the code from the translator pages, and I couldn't find any new languages.
I tested some of the new languages and some work, some don't, and some are just aliases like `es-MX` or `en-GB`. This can be determined because the API returns the language codes as `es` and `en`. Please make sure all the new languages are working and that their `NativeName` properties contain the native name and not the English name. Preferably use Wikipedia to get the native names (for consistency with the other languages).

> Class to allow external exposure to the language list.

The language list can already be accessed through the `Language.LanguageDictionary` property. It's a static property so the list is only created once.

> The language list is sorted, so as to more easily identify any duplicates, and it makes it easier to find a language on the list.

The language list was already sorted by language name but I'm OK with the new order.

> I added an extra try catch on AggregateTranslator.cs, because in some cases an exception was thrown inside the catch code.

There's no way an exception could be thrown there. The code is just adding `Exception` objects to the list. Could you provide a sample code that produces an exception there?

David-Maisonave committed:
Update LanguageDictionary.cs

David-Maisonave commented:
Hi,

> The extra comments are unnecessary too.

I'll remove the extra comments.

> Could you please edit the code so the only changes are made to the _languages dictionary? 

I made the change to AggregateTranslator.cs, because the logic adding an exception to a list was throwing an exception. It kept happening when http error 429 would occur.  If you're sure you don't want that change, I can remove until you pull the changes.

I see the LanguageServiceDetails class as an essential class to the GTranslate library.  There has to be a way for any program using GTranslate to be able to tell what languages are supported, and which language tags to use.
By exposing the supported language details, there's no need to have the "alias" logic, because the calling application doesn't have to just blindly send a request.  Moreover, with my testing, this alias logic cause multiple problems, until I got rid of it.

> Also, you can undo the changes to the other files through your code editor.

I'm going to use my branch of GTranslate to keep the changes that don't make it into the main branch, because I need these changes for my project.
If you're not able to select which files to take, I'll temporarily remove the changes in the other files until you pull in the changes.
I can also temporarily put back the alias code as well.

> Could you please provide your sources for these new languages?

### I used the name of the languages from the following link:
https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes


### Individual Translating Service Source:
* https://developers.google.com/admin-sdk/directory/v1/languages
* https://docs.microsoft.com/en-us/azure/cognitive-services/translator/language-support
* https://cloud.yandex.com/en/docs/translate/concepts/supported-languages
* https://sites.google.com/site/opti365/translate_codes

### Recently Google added 24 languages:
* https://www.blog.google/products/translate/24-new-languages/
* https://ai.googleblog.com/2022/05/24-new-languages-google-translate.html
* These new languages are also listed on Google Translate online interface:
  * https://translate.google.com/

However, I could not find a good list showing what language tags they're using for the new languages. I was only able to get a little over half of them to work.

### Microsoft Service Pack and Language Interface Pack
* https://docs.microsoft.com/en-us/windows-hardware/manufacture/desktop/available-language-packs-for-windows?view=windows-11
* These list had some languages which were not in the list.
* I was able to add most of the missing languages.
* I believe Cherokee is the only language I could not get to work, which should be supported by Bing or Microsoft since they have it on the Language Interface Pack.  
  * But I notice that Cherokee is **NOT** in the drop down list when using the online interface.
  * https://www.bing.com/translator

FYI:
I lived in the Philippines for 4 years (1983-1987) when I was station in Clark AB Philippines. And back then, a person saying that the locals spoke Filipino was like saying Mexicans spoke Mexican. So I was a little annoyed when I saw Filipino listed in MS Language Pack and on Bing Translator.
But I did very recently do a google search on Tagalog v.s. Filipino, and much to my surprise they consider them two different languages.
https://learningfilipino.com/blog/difference-between-tagalog-and-filipino/

However, Filipino is derived from Tagalog.

With that said, I think it's a good idea to keep both languages on the list, but with each one referencing the other.

David-Maisonave committed:
Update LanguageDictionary.cs

David-Maisonave commented:
Since last post was very long, I'm answering in multiple post.

> I tested some of the new languages and some work, some don't, and some are just aliases like es-MX or en-GB. This can be determined because the API returns the language codes as es and en.

Please keep in mind that I removed the alias logic when testing.  In my code, I'm never going to use the alias logic, and IMHO, it should be removed, or the API should have an option to bypass the alias logic. Every time a new language is added, the alias logic well always be a concern.
I tested all the languages that I added, and they worked for me.  I also speak Spanish, so it was easy to verify es-MX.

Can you give me some details on what's not working.
* What target language?
* What's the source text?
* What's the source language?
* How did it fail?

I recommend you try the [ABetterTranslator](https://github.com/David-Maisonave/ABetterTranslator). It has an installer, and there's a tab you can use to test an individual language.  It works on the fly, while the user is typing in the text.
I'm going to update the program today, with the new version having a feature to save a set of languages.

> Please make sure all the new languages are working and that their NativeName properties contain the native name and not the English name. Preferably use Wikipedia to get the native names (for consistency with the other languages).

I didn't test the languages using either the English name nor the native name.  My code only uses the language tag when sending a translation request.
The English names, and the native names are only used in the GUI.

> The language list can already be accessed through the Language.LanguageDictionary property. It's a static property so the list is only created once.

I did **not** realize that!I've taken out that class and associated logic.

I'm mainly a C++ programmer, so there's some aspects to C# which are still new to me. Static classes is something I just very recently learn while trying to address a different issue.

> There's no way an exception could be thrown there. The code is just adding Exception objects to the list. Could you provide a sample code that produces an exception there?

I also thought the same thing. I believe some code corruption is triggering the issue. I'll see if I can reproduce the issue, and get more details.

David-Maisonave commented:
Sorry, but there's something I should have mention in my last post.
I used the ABetterTranslator program on the program itself to make it multilingual. It was working just find until yesterday when more then a dozen new strings were added.  That gave it a total of 200 strings to translate.  At that point 22 languages failed to translate. These 22 languages include languages that were already in GTranslate.

The program still works on a Resx file that has 77 strings, so I believe the failure has to do with the quantity of strings and the assigned translator.

What is common about all the failed languages, is that they're all the languages which explicitly only has Google as the translator.
If you look at the below screenshot, all the languages except for Somali are listed as having only Google as the translator.
<img width="449" alt="The22FailedGoogleOnlyLanguages" src="https://user-images.githubusercontent.com/47364845/187799655-bb8f714e-fed3-44b6-8624-00ad19cace55.png">

Now,  yesterday Somali was failing, and it only had Google as the translator.  Today, I merged the code with the latest changes on GTranslate, which change Somali to having the multiple translator.

After that change, Somali now works.

I originally thought the Google Translator was failing, but now I think there has to be something else within the GTranslate code that is triggering the issue.

Corsican is the first language alphabetically that was already in GTranslate, before my changes. So I tried to reproduce the issue with that language.  But I can't seem to reproduce the issue manually.
Here's the error I get when it runs through the normal process:
`GTranslate_..:toLang(co):fromLang(en):Translator=(AggregateTranslator): No translator provided a valid result. (Response status code does not indicate success: 429 (Too Many Requests).) (Unable to get the data from the response.)`


**Is there anything unique about languages having only Google Translate that could trigger this issue?**

d4n3436 commented:
> Please keep in mind that I removed the alias logic when testing. In my code, I'm never going to use the alias logic, and IMHO, it should be removed, or the API should have an option to bypass the alias logic.

The dictionary of aliases was created to make GTranslate easier to use by recognizing that a specific string belongs to a language name, alternative name or native name and then obtaining its ISO code. Removing the dictionary of aliases would be bad for usability and performance because instead of making 2 lookups the code would have to check all entries in the dictionary and search for a match in 3 properties.

> I tested all the languages that I added, and they worked for me. I also speak Spanish, so it was easy to verify es-MX.

I know, but I was referring to the fact that the API itself returns an `es` code when translating text to `es-MX` indicating that the API treats `es-MX` and `es` the same, thus making the addition of `es-MX` unnecessary.

> I didn't test the languages using either the English name nor the native name. My code only uses the language tag when sending a translation request.
The English names, and the native names are only used in the GUI.

I was referring to update the new languages to have their `NativeName` properties actually be the native names and not the English name.

An example:
```cs
["sa"] = new("Sanskrit", "Sanskrit", "sa", "san", TranslationServices.Google),
```
The second parameter should be the native name, which is "संस्कृतम्".

> Is there anything unique about languages having only Google Translate that could trigger this issue?

That's happening because Google is IP banning you for making too many requests in a short period of time. The 429 error code reflects that, and that affects both `GoogleTranslator` and `GoogleTranslator2` classes. Since all available translators have thrown an exception, the `AggregateTranslator` will throw an `AggregateException` because it can't continue.

The second exception (`Unable to get the data from the response.`) comes from `GoogleTranslator2`. That happens when the translator can't parse the response because the string that is supposed to contain the (actual) translation response is null.

David-Maisonave commented:
> I know, but I was referring to the fact that the API itself returns an es code when translating text to es-MX indicating that the API treats es-MX and es the same, thus making the addition of es-MX unnecessary.

How would the API do this without the alias code?

> The second parameter should be the native name, which is "संस्कृतम्".

I understand now. But where are you getting the native names from? I did a quick google search, and I didn't see a list of ISO 639-1 having the native names.

> The second exception (Unable to get the data from the response.) comes from GoogleTranslator2. 

When does the code use GoogleTranslator v.s. GoogleTranslator2?

> That's happening because Google is IP banning you for making too many requests in a short period of time. 

That's what I originally thought. But that wouldn't explain why **Somali** failed when only Google service was assigned, and ran successfully when Bing and Microsoft was added.

* In other words, the following code fails:
  * `["so"] = new("Somali", "Af Soomaali", "so", "som", TranslationServices.Google ),`

* And this code works:
  * `["so"] = new("Somali", "Af Soomaali", "so", "som", TranslationServices.Google | TranslationServices.Bing | TranslationServices.Microsoft),`

That is unless the code is setup to jump to another translation service when one service fails.  Maybe I missed it, but I didn't see that type of logic in GTranslate.

In any case, I did find a workaround for this error.  I had an upper limit of 10,000 characters per translation request.  Every language started working after I lower the upper limit to 5,000.
The code still translates the same amount of data, but when the limit is reached, it breaks up the translation into multiple translation requests.  It now makes twice as many translation request with few characters in each request.

This further proves that the issue is not related to having too many request, and instead it's being triggered  because of the amount of characters in a single translation request.

But I'm still not sure why 10,000 is a good limit for almost all the languages except for those only having Google as the translating service.

David-Maisonave commented:
I did find a link with native names:
https://omniglot.com/language/names.htm
And it has an [Excel spreadsheet](https://omniglot.com/charts/languagenames.xls) download link. 

I'll update the code with the native names.

David-Maisonave committed:
Update LanguageDictionary.cs

David-Maisonave committed:
Update LanguageDictionary.cs

d4n3436 commented:
> How would the API do this without the alias code?

I don't understand what you mean.

> When does the code use GoogleTranslator v.s. GoogleTranslator2?
> That is unless the code is setup to jump to another translation service when one service fails.

That's what `AggregateTranslator` does. It uses the first available translator and if that translator throws an exception then it uses the next one and so on. If there are no translators left, it throws an `AggregateException` containing all the exceptions.

You can see the order of the translators in AggregateTranslator.cs:

```cs
public AggregateTranslator()
    : this(new GoogleTranslator(), new GoogleTranslator2(), new MicrosoftTranslator(), new YandexTranslator(), new BingTranslator())
{
}
```

> In any case, I did find a workaround for this error. I had an upper limit of 10,000 characters per translation request. Every language started working after I lower the upper limit to 5,000.

Both GoogleTranslator and GoogleTranslator have a limit of 5000 characters. If that limit is excceded then the API truncates the text.
YandexTranslator has a limit of 10000 characters.
BingTranslator has a limit of 1000 characters and MicrosoftTranslator has a limit of 50000 characters.

d4n3436 commented on README.md:
```diff
@@ -1,24 +1,7 @@
-- An `AggregateTranslator` class that groups the default translators for ease of use with the ability to add custom translators.
-
-See the [examples](examples) folder for an example program.
+These changes were made inorder to be used in project [ABetterTranslator](https://github.com/David-Maisonave/ABetterTranslator).
```
Please undo these changes.

David-Maisonave replied:
done.

d4n3436 commented on src/GTranslate/GTranslate.csproj:
```diff
@@ -18,13 +18,16 @@
+  </PropertyGroup>
+  <PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Release|net7.0-windows|AnyCPU'">
+    <NoWarn>1701;1702;</NoWarn>
+  </PropertyGroup>
```
Please undo these changes.

David-Maisonave replied:
done

d4n3436 commented on src/GTranslate/Translators/AggregateTranslator.cs:
```diff
@@ -99,9 +99,17 @@ public async Task<ITranslationResult> TranslateAsync(string text, string toLangu
+                    throw new Exception($"Error while processing exception. New Exception={ee.Message}; Original Exception={e.Message}");
+                }
+            }
+            // Consider adding sleep here for http error 429
```
Please undo these changes.

David-Maisonave replied:
done

d4n3436 commented on src/GTranslate/LanguageDictionary.cs:
```diff
@@ -142,153 +142,167 @@ public bool TryGetLanguage(string code, [MaybeNullWhen(false)] out Language lang
         ["no"] = new("Norwegian", "Norsk", "no", "nor"),
+        ["nso"] = new("Sepedi", "Sesotho sa Leboa", "nso", "nso", TranslationServices.Google),
+        ["ny"] = new("Chichewa", "Nyanja", "ny", "nya", TranslationServices.Google),
+        ["om"] = new("Oromo", "Afaan Oromoo [ʌfɑːn ɔrɔmoː] Oromiff", "om", "orm", TranslationServices.Google),
```
The native language is just "Afaan Oromoo".
```suggestion
        ["om"] = new("Oromo", "Afaan Oromoo", "om", "orm", TranslationServices.Google),
```

David-Maisonave replied:
Fixed

d4n3436 commented on src/GTranslate/LanguageDictionary.cs:
```diff
@@ -142,153 +142,167 @@ public bool TryGetLanguage(string code, [MaybeNullWhen(false)] out Language lang
+        ["ps"] = new("Pashto", "پښتو", "ps", "pus", TranslationServices.Google | TranslationServices.Bing | TranslationServices.Microsoft),
         ["pt"] = new("Portuguese", "Português", "pt", "por"),
-        ["pa"] = new("Punjabi", "ਪੰਜਾਬੀ", "pa", "pan"),
+        ["pt-BR"] = new("Portuguese (Brazil)", "Português (Brazil)", "pt-BR", "pt-BR", TranslationServices.Google),
```
As mentioned in the comments, The APIs don't treat `pt` and `pt-BR` as separate languages, but Microsoft Azure actually does with `pt-PT`.
```suggestion
        ["pt-PT"] = new("Portuguese (Portugal)", "Português (Portugal)", "pt-PT", "pt-PT", TranslationServices.Bing | TranslationServices.Microsoft),
```

David-Maisonave replied:
> As mentioned in the comments, The APIs don't treat pt and pt-BR as separate languages, but Microsoft Azure actually does with pt-PT.

I'm sorry, but I still don't understand this.  I mention my confusion on this in previous post, but I didn't get a clear answer.

* It sounds like you're saying the API ignores what's after the dash.  But if that was the case, zh-CN and zh-TW would return the same results.  But they don't.  I get two separate results.  Exactly how does the API treat pt and pt-BR the same?
* When Microsoft Azure treats it as separate languages, does it produce different translation results for pt vs pt-PT?

Can you please provide more details?
Thanks

d4n3436 replied:
> It sounds like you're saying the API ignores what's after the dash.

No, the APIs have a list of languages that can be used to translate. In Google Translate you can either see it in their official documentation and through the source code of translate.google.com.
The official documentation displays 109 supported languages and in translate.google.com you can see 133 supported languages. Those 2 sites display the 2 variants of chinese `zh-CN` and `zh-TW` but there's only one language code for Portuguese (`pt`).

[Google Translate](https://translate.google.com/):
![image](https://user-images.githubusercontent.com/39978852/188332029-816e4813-10e9-4f62-aa17-05c9ee1c6d8b.png)

[Google Cloud Translation docs](https://cloud.google.com/translate/docs/languages):
![image](https://user-images.githubusercontent.com/39978852/188332055-527fdae2-3020-435c-8bbf-482709928230.png)

This is different in Bing/Microsoft Azure. Both Bing Translator and Microsoft Azure Translator display separate language codes for the two Portuguese variants.

[Bing Translator](https://www.bing.com/translator):
![image](https://user-images.githubusercontent.com/39978852/188332152-161a0ea6-cd11-4bed-a86e-d00cf5e5931f.png)

[Microsoft Azure Translator docs](https://docs.microsoft.com/en-us/azure/cognitive-services/translator/language-support):
![image](https://user-images.githubusercontent.com/39978852/188332169-cd61ba6e-81f4-4a72-ac93-dbe28d496c01.png)

David-Maisonave replied:
I don't understand why Microsoft would default pt to Portuguese (Brazil).  Portuguese originated in Portugal, so that should be the default Portuguese for pt.

I've reverted it back.

Done.

d4n3436 commented on src/GTranslate/LanguageDictionary.cs:
```diff
@@ -142,153 +142,167 @@ public bool TryGetLanguage(string code, [MaybeNullWhen(false)] out Language lang
+        ["en-GB"] = new("English (UK)", "English (UK)", "en-GB", "en-GB", TranslationServices.Google),
         ["eo"] = new("Esperanto", "Esperanto", "eo", "epo", TranslationServices.Google | TranslationServices.Yandex),
+        ["es"] = new("Spanish", "Español", "es", "spa"),
+        ["es-MX"] = new("Spanish (Mexico)", "español (México)", "es-MX", "es-MX", TranslationServices.Bing | TranslationServices.Microsoft),
```
As mentioned in the comments, this should be removed because the APIs treat `es` and `es-MX` as the same language.
```suggestion
```

David-Maisonave replied:
Again, can you please provide more details as to how es is treated as the same as es-MX?

I just ran a test to translate the word "**Computer**".
* In "es" I got the following results:
  * Computadora
* In "es-MX" I got the following results:
  * Ordenador

This is using GTranslate to translate the word.  Unless you know the language differences, the translations can appear identical because 99% of the translations will be the same.  But a few words are different.

d4n3436 replied:
That's happening because the result from `es` doesn't come from Bing/Microsoft translator.

If you modify line 66 of AggregateTranslator.cs so that the only translator present is `MicrosoftTranslator`..
```diff
- _translators = new ITranslator[] { googleTranslator, googleTranslator2, microsoftTranslator, yandexTranslator, bingTranslator };
+ _translators = new ITranslator[] { microsoftTranslator };
```
..and then use `AggregateTranslator` through the example program, you will notice that both `es` and `es-MX` return the same result.

![image](https://user-images.githubusercontent.com/39978852/188334212-20c52a0a-1c80-4134-99bd-36941d83837b.png)

If you try to do this with `BingTranslator` you'll get a 400 Bad Request status code.

David-Maisonave replied:
From my test, the correct translation is coming from TranslationServices.Microsoft.

TranslationServices.Bing fails when it's the only one listed as a translation service.  Which means the correct change should be the following:
`["es-MX"] = new("Spanish (Mexico)", "español (México)", "es-MX", "es-MX", TranslationServices.Microsoft),`

If it works with above values, why remove it?

d4n3436 replied:
> If it works with above values, why remove it?

Because it's misleading, there isn't a `es-MX` language code listed in the docs, and the API treats `es` and `es-MX` as the same language because the API returns an `es` target language as you can see in the image above. You can even get the same result by creating a fake language code like `es-LL` or `en-KO`. This means the API just disregards anything after the "-" when it's not required.

David-Maisonave replied:
> Because it's misleading, there isn't a es-MX language code listed in the docs, and the API treats es and es-MX as the same language because the API returns an es

I was confused before because when you said API, I thought you were referring to GTranslate API.  From above comment, it seems you're referring to the translation service API.

With that said, I thought you're saying it's misleading because the translation service returns "es" in its response. But when I tested this out, I don't see "es" by itself any where in the response returned by GTranslate API.

Here's the code calling GTranslate API:
`GTranslate.Results.ITranslationResult result = await translator.TranslateAsync("Computer", "es-MX", "en");`

Here's the content of the result variable:
<img width="783" alt="es-MX_Result" src="https://user-images.githubusercontent.com/47364845/188486127-b895480c-e276-4828-86f7-c7c0c4926c5c.png">

GTranslate API results are **NOT** misleading with the input vs the output.

d4n3436 replied:
That's easy to explain. In line 219 of MicrosoftTranslator.cs you'll notice that the method returns `Language.GetLanguage(toLanguage.ISO6391)`. This means the method ignores the target language returned from the API response because I expected Microsoft Translator would return an error code when an invalid language code is passed instead of just cutting the extra part.

If you make these changes you'll get the correct target language:
```diff
+ string targetLanguage = root.GetProperty("translations")[0].GetProperty("to").GetString();
- return new MicrosoftTranslationResult(translation, text, Language.GetLanguage(toLanguage.ISO6391), Language.GetLanguage(sourceLanguage), score);
+ return new MicrosoftTranslationResult(translation, text, Language.GetLanguage(targetLanguage), Language.GetLanguage(sourceLanguage), score);
```

David-Maisonave replied:
> This means the method ignores the target language returned from the API response 

That sounds good. That means the code will work as-is with es-MX.

I still don't understand the justification for removing it, if it fully works.

d4n3436 replied:
I already mentioned why [here](https://github.com/d4n3436/GTranslate/pull/3#discussion_r962540851).

d4n3436 commented on src/GTranslate/LanguageDictionary.cs:
```diff
@@ -142,153 +142,167 @@ public bool TryGetLanguage(string code, [MaybeNullWhen(false)] out Language lang
+        ["el"] = new("Greek", "Ελληνικά", "el", "ell"),
+        ["emj"] = new("Emoji", "Emoji", "emj", "emj", TranslationServices.Yandex), // Not present in Yandex.Cloud
         ["en"] = new("English", "English", "en", "eng"),
+        ["en-GB"] = new("English (UK)", "English (UK)", "en-GB", "en-GB", TranslationServices.Google),
```
As mentioned in the comments, this should be removed because the API treats `en` and `en-GB` as the same language.
```suggestion
```

David-Maisonave replied:
Here, it does seem that en-GB is ignored, and all the translations are coming back in en-US.
I tried entering a British sentence in the online [Google ](https://translate.google.com/?sl=auto&tl=en&text=The%20honour%20of%20the%20barrister%20was%20attacked%20by%20his%20neighbour%20who%20had%20a%20biscuit%20in%20the%20boot%20of%20his%20car%20due%20to%20his%20good%20humour%20and%20hard%20labour.%20&op=translate) and [Bing  ](https://www.bing.com/translator) translator, and both brought back the exact same sentence.

> The honour of the barrister was attacked by his neighbour who had a biscuit in the boot of his car due to his good humour and hard labour. 

The above sentence should have come back to the following:
>The honor of the attorney was attacked by his neighbor who had a cookie in the trunk of his car due to his good humor and hard labor. 

Even the following link, only changed four words (honour, neighbour, humour, & labour).
https://gotranscript.com/translation-services/british-to-american#text

It doesn't look like there is a good translation service for en-GB.

I've removed it.

Done.

d4n3436 commented on src/GTranslate/LanguageDictionary.cs:
```diff
@@ -142,153 +142,167 @@ public bool TryGetLanguage(string code, [MaybeNullWhen(false)] out Language lang
         ["da"] = new("Danish", "Dansk", "da", "dan"),
-        ["nl"] = new("Dutch", "Nederlands", "nl", "nld"),
+        ["de"] = new("German", "Deutsch", "de", "deu"),
+        ["dgo"] = new("Dogri", "डोगरी (ḍogrī) ڈوگرى", "dgo", "dgo", TranslationServices.Google),
```
The native name should be just "डोगरी".
```suggestion
        ["dgo"] = new("Dogri", "डोगरी", "dgo", "dgo", TranslationServices.Google),
```

David-Maisonave replied:
Fixed.

d4n3436 commented on src/GTranslate/LanguageDictionary.cs:
```diff
@@ -142,153 +142,167 @@ public bool TryGetLanguage(string code, [MaybeNullWhen(false)] out Language lang
         ["ar"] = new("Arabic", "العربية", "ar", "ara"),
-        ["hy"] = new("Armenian", "Հայերեն", "hy", "hye"),
+        ["as"] = new("Assamese", "অসমীয়া", "as", "asm", TranslationServices.Bing | TranslationServices.Microsoft),
+        ["ay"] = new("Aymara", "aymar aru [ajmaɾ ˈaɾu]", "ay", "aym", TranslationServices.Google),
```
Native name should be "Aymar aru".
```suggestion
        ["ay"] = new("Aymara", "Aymar aru", "ay", "aym", TranslationServices.Google),
```

David-Maisonave replied:
fixed

d4n3436 commented on src/GTranslate/LanguageDictionary.cs:
```diff
@@ -142,153 +142,167 @@ public bool TryGetLanguage(string code, [MaybeNullWhen(false)] out Language lang
         ["pt"] = new("Portuguese", "Português", "pt", "por"),
-        ["pa"] = new("Punjabi", "ਪੰਜਾਬੀ", "pa", "pan"),
+        ["pt-BR"] = new("Portuguese (Brazil)", "Português (Brazil)", "pt-BR", "pt-BR", TranslationServices.Google),
+        ["qu"] = new("Quechua", "Kechua", "qu", "que", TranslationServices.Google),
```
The native name should be "Runa simi".
```suggestion
        ["qu"] = new("Quechua", "Runa simi", "qu", "que", TranslationServices.Google),
```

David-Maisonave replied:
https://omniglot.com/language/names.htm
The above link shows Kechua as primary, and Runa simi as secondary. And to make it more complicated, when I translate the word "Quechua" to English, I get "Runasimi".

IMHO, the _languages shouldn't have the **native name** included. Instead this should be done programmatically. Why do something manually when the program can do this for you?

I've changed it to "Runa simi".

Done.

d4n3436 replied:
> IMHO, the _languages shouldn't have the native name included. Instead this should be done programmatically. Why do something manually when the program can do this for you?

What do you mean? That the language list should be source-generated? That's something I'm planning to do.
Right now I can scrape the language list from the source code of Google Translate's page. This includes the English name and native name too. With Bing/Microsoft Azure this is easier as there's an API endpoint that provides that info. Unfortunately it's not possible to get the ISO 639-3 code from these methods but I guess there's a API somewhere that would provide that info.

d4n3436 commented on src/GTranslate/LanguageDictionary.cs:
```diff
@@ -142,153 +142,167 @@ public bool TryGetLanguage(string code, [MaybeNullWhen(false)] out Language lang
+        ["eu"] = new("Basque", "Euskara", "eu", "eus"),
+        ["fa"] = new("Persian", "فارسی", "fa", "fas"),
         ["fi"] = new("Finnish", "Suomi", "fi", "fin"),
+        ["fil"] = new("Filipino", "Tagalog", "fil", "fil", TranslationServices.Bing | TranslationServices.Microsoft),
```
Isn't the native name "Filipino"?

David-Maisonave replied:
No.

David-Maisonave replied:
Did you read the link I posted?

https://learningfilipino.com/blog/difference-between-tagalog-and-filipino/

Although Tagalog and Filipino are technically supposed to be separate languages, many (including Pilipino's) see them as one and the same.

Alternatively, **Pilipino** could be used as the native name.
> On the other hand, Pilipino, is how the locals from the Philippines refer to themselves, or to their national language.

Above quote comes from the following link:
http://www.hawaii.edu/cps/filipino.html

d4n3436 replied:
I know, but I was referring to how Filipinos refer to its name in their language. From Wikipedia's page it's referred as just "Filipino" or "Wikang Filipino".

David-Maisonave replied:
I've changed it to Pilipino.

David-Maisonave replied:
> I know, but I was referring to how Filipinos refer to its name in their language. From Wikipedia's page it's referred as just "Filipino" or "Wikang Filipino".

In the wiki, do keyword search for "Pilipino".
You'll find it in Further history.  The reason why the native name should not be Filipino is because most Pilipino languages do not have the "F" sound.

Filipino is the anglicize version of the name.

d4n3436 replied:
From Wikipedia:

> A compromise was reached and the wording on the 1973 constitution made no mention of dropping the national language Pilipino or made any mention of Tagalog. Instead, the 1973 Constitution, in both its original form and as amended in 1976, designated English and Pilipino as official languages and provided for development and formal adoption of a common national language, termed Filipino, to replace Pilipino.

From https://learningfilipino.com/blog/difference-between-tagalog-and-filipino/:

> You see Tagalog was one of the native languages of the Philippines. It was a native language long before the Spaniards and Americans colonized the Philippines. At some point, Filipinos wanted a native language to become the national language of the Philippines.  And Tagalog was the chosen native language to become the basis for this national language. Tagalog was then renamed “Pilipino” and eventually “Filipino.”

> Renaming of Pilipino to Filipino

> By 1987, “Pilipino” was renamed “Filipino”, the one letter change from “p” to “f” marking a significant shift in the Filipino language. The 20-letter Tagalog abakada was replaced with the 28-letter [Filipino alphabet](https://learningfilipino.com/blog/language/filipino-alphabet/), and included letters from the English and Spanish alphabet not present in the former ABaKaDa alphabet.


It seems that at some point Filipino was named Pilipino but in 1987 it was changed to Filipino.


Even in [Wikipedia in Tagalog](https://tl.wikipedia.org/wiki/Wikang_Filipino) Filipino is referred as just "Filipino".

I took the native names of the languages from Wikipedia and from Microsoft's own API endpoint for languages (https://api.cognitive.microsofttranslator.com/languages?api-version=3.0&scope=translation).
In both places the native name of Filipino is "Filipino" so I don't think Wikipedia and Microsoft would have wrong language data.

![image](https://user-images.githubusercontent.com/39978852/188395278-4a0bd23b-65dd-459f-93db-9e75b51f0efa.png)

David-Maisonave replied:
> I took the native names of the languages from Wikipedia and from Microsoft's own API

Considering how they're handling "pt" as PT-Brazil, I wouldn't consider them an expert on the matter.

The proceeding quote comes from the following article: 
[https://learningtagalog.com/articles/tagalog_filipino_pilipino_difference.html](https://learningtagalog.com/articles/tagalog_filipino_pilipino_difference.html)
> "Filipino is still the same as Tagalog, which unless otherwise specified, is generally understood as standard or Manila Tagalog. The grammar is identical."


In any case, I give up. I'll change it to Filipino until after the update.

d4n3436 replied:
> Considering how they're handling "pt" as PT-Brazil, I wouldn't consider them an expert on the matter.

Every translator treats it like that, and that's because Brazilian portuguese is the most common variant of the two.

d4n3436 commented on src/GTranslate/LanguageDictionary.cs:
```diff
@@ -142,153 +142,167 @@ public bool TryGetLanguage(string code, [MaybeNullWhen(false)] out Language lang
+        ["fr-CA"] = new("French (Canada)", "Français (Canada)", "fr-CA", "fr-CA", TranslationServices.Bing | TranslationServices.Microsoft),
         ["fy"] = new("Frisian", "Frysk", "fy", "fry", TranslationServices.Google),
+        ["ga"] = new("Irish", "Gaeilge", "ga", "gle"),
+        ["gd"] = new("Gaelic", "Gàidhlig", "gd", "gla", TranslationServices.Google | TranslationServices.Yandex),
```
This language is better known as "Scottish Gaelic".

David-Maisonave replied:
[ISO 639-1](https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes) list the primary language as Gaelic.

d4n3436 replied:
Yeah but the language dictionary had the language name previously listed as "Scottish Gaelic". Changing it to "Gaelic" would break the code of people that rely on the previous name. A solution would be to add "Scottish Gaelic" as an alias.

David-Maisonave replied:
Google Translate list it as "Scots Gaelic".
So I changed it back to "Scottish Gaelic"

Done.

d4n3436 commented on src/GTranslate/LanguageDictionary.cs:
```diff
@@ -142,153 +142,167 @@ public bool TryGetLanguage(string code, [MaybeNullWhen(false)] out Language lang
-        ["tr"] = new("Turkish", "Türkçe", "tr", "tur"),
+        ["ti"] = new("Tigrinya", "ትግር", "ti", "tir", TranslationServices.Bing | TranslationServices.Microsoft),
         ["tk"] = new("Turkmen", "Türkmen Dili", "tk", "tuk", TranslationServices.Google | TranslationServices.Bing | TranslationServices.Microsoft),
+        ["tl"] = new("Tagalog", "Filipino", "tl", "tgl", TranslationServices.Google | TranslationServices.Yandex),
```
The native name should be "Tagalog".
```suggestion
        ["tl"] = new("Tagalog", "Tagalog", "tl", "tgl", TranslationServices.Google | TranslationServices.Yandex),
```

David-Maisonave replied:
I'll change this to Pilipino for the native language.
See previous comment.

d4n3436 replied:
I'm a bit confused because the [page you provided](https://omniglot.com/language/names.htm) for the native names say Tagalog's native name is just Tagalog or Wikang Tagalog. The same can be seen on Wikipedia's page.
![image](https://user-images.githubusercontent.com/39978852/188333721-8ea64add-a752-47c1-9f22-c765c6b64508.png)

David-Maisonave replied:
>   On the other hand, Pilipino, is how the locals from the Philippines refer to themselves, or to their national language.

The above quote comes from the following link:
http://www.hawaii.edu/cps/filipino.html

d4n3436 replied:
Yeah but that page it's not talking about Tagalog but rather its standarized form, Filipino. What I'm asking is, what's the endonym/native name of "Tagalog"? How does people refer to "Tagalog" in Tagalog? because I'm pretty sure it's not "Pilipino".

David-Maisonave replied:
> How does people refer to "Tagalog" in Tagalog? because I'm pretty sure it's not "Pilipino".

It most certainly is referred to as Pilipino.

As I stated previously, I lived in the Philippines for 4 years (1983-1987).

I considered that my experience might be out dated, but the following links shows it still to be true.
* [An Introduction to Pilipino Phonetics and Pronunciation](https://www.karger.com/Article/Pdf/258574)
* ["Pilipino" is how we, locals from the Philippines, refer to ourselves, or the national language](https://hinative.com/questions/8148444)
* [Pilipino, is how the locals from the Philippines refer to themselves, or to their national language](http://www.hawaii.edu/cps/filipino.html)
* [Filipino and Pilipino are simply different names for Tagalog](https://learningtagalog.com/articles/tagalog_filipino_pilipino_difference.html)

d4n3436 commented on src/GTranslate/LanguageDictionary.cs:
```diff
@@ -142,153 +142,167 @@ public bool TryGetLanguage(string code, [MaybeNullWhen(false)] out Language lang
+        ["tw"] = new("Twi", "Twi", "tw", "twi", TranslationServices.Google),
+        ["ty"] = new("Tahitian", "Reo Tahiti", "ty", "tah", TranslationServices.Bing | TranslationServices.Microsoft),
+        ["udm"] = new("Udmurt", "Удмурт кыл", "udm", "udm", TranslationServices.Yandex),
+        ["ug"] = new("Uighur", "ئۇيغۇرچە", "ug", "uig", TranslationServices.Google | TranslationServices.Bing | TranslationServices.Microsoft),
```
This should be changed back to "Uyghur" as it's the preferred spelling.
```suggestion
        ["ug"] = new("Uyghur", "ئۇيغۇرچە", "ug", "uig", TranslationServices.Google | TranslationServices.Bing | TranslationServices.Microsoft),
```

David-Maisonave replied:
[ISO 639-1](https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes) list the primary language name as Uighur and it has Uyghur as the secondary name.

I Googled this, and your right, in that now the preferred spelling is Uyghur.

I've corrected it.

Done.

d4n3436 replied:
Same as Scottish Gaelic. A solution would be to add "Uyghur" as an alias.

d4n3436 reviewed (changes requested):
Last reviews. After testing the new languages it should be ready to merge.

David-Maisonave committed:
Update README.md

David-Maisonave committed:
Update GTranslate.csproj

David-Maisonave committed:
Update AggregateTranslator.cs

David-Maisonave committed:
Update LanguageDictionary.cs

David-Maisonave committed:
Update LanguageDictionary.cs

David-Maisonave committed:
Update LanguageDictionary.cs

David-Maisonave committed:
Update LanguageDictionary.cs

David-Maisonave committed:
Update LanguageDictionary.cs

David-Maisonave committed:
Update LanguageDictionary.cs

David-Maisonave committed:
Update LanguageDictionary.cs

David-Maisonave committed:
Per request changed Tagalog and Filipino native name

David-Maisonave commented:
The following line has the wrong label:
` ["zh-TW"] = new("Chinese (Traditional)", "繁體中文 (繁體)", "zh-TW", "zho-TW", TranslationServices.Google | TranslationServices.Bing | TranslationServices.Microsoft),
`
That should be Chinese (Traditional, Taiwan) or Chinese (Taiwan).
That was bothering me for a while, but I didn't know what was the correct code for Chinese (Traditional) until you posted the following link: https://api.cognitive.microsofttranslator.com/languages?api-version=3.0&scope=translation

That line should be changed to the following two lines:
```
        ["zh-Hant"] = new("Chinese (Traditional)", "繁體中文 (繁體)", "zh-Hant", "zh-Hant",  TranslationServices.Microsoft),
        ["zh-TW"] = new("Chinese (Taiwan)", "中文（臺灣", "zh-TW", "zho-TW", TranslationServices.Google | TranslationServices.Bing | TranslationServices.Microsoft),
```

I ran a test, and all three of the languages selected in below snapshot give different results per their associated language.
<img width="364" alt="ChineseTraditional" src="https://user-images.githubusercontent.com/47364845/188497097-f349ce26-0ef4-489b-85f6-7c1f03fe064d.png">

d4n3436 commented:
`zh-Hant` and `zh-Hans` are ISO-15924 language codes that only `BingTranslator`/`MicrosoftTranslator` use.
`zh-Hant` is equivalent to `zh-TW` and `zh-Hans` is equivalent to `zh-CN` and there's already aliases for them so they redirect to `zh-TW` and `zh-CN`.

David-Maisonave commented:
FYI:
Just to give you a heads-up, I plan to have my own variation of GTranslate which will have languages that don't make it on this update.
I'm also going to add the following changes.
* Moving the language details out of the source code and into either an XML file or a SQLite DB.
  * That will make it so the source code doesn't have to be changed every time there are additional languages.
* Have the language details only include the key and language name.
  * Remove native-name because it's not needed to perform the translation.
  * Remove iso6391 because it's misleading since many languages don't have an ISO639-1 2 letter tag.
  * Remove iso6393 because it's misleading since some languages don't have an ISO639-3 3 letter tag.
* Removal much of the GTranslate alias language code.
  * Keep only tag alias that are exact match.  Example: ["sr-Cyrl"] = "sr"
* Move the remaining alias code to XML file. 
  * That will allow developers to include or exclude what they want from the alias list.
* Add additional try-catch code.
* Use SQLite DB to cache translations. 
  * The DB will store strings that were previously translated, so-as to avoid calling translation service for the same string and target language.
* Add logic to handle translations which exceed the character limit for the translation service, by dividing the translation request into multiple request which are below the limit.

d4n3436 commented:
This is FOSS software, feel free to fork the project and adjust it to your needs. Be sure to follow the conditions of the MIT license.

This would've been merged much faster if it wasn't for the changes to existing languages and the addition of non-existing/redundant languages, but I appreciate the contribution nonetheless.
